azure/src/main/scala/quasar/blobstore/azure/Azure.scala

```diff
@@ -17,43 +17,98 @@
+        rx.publisherToStream[F, AccessToken](
+          (new ClientSecretCredentialBuilder)
+            .clientId(clientId.value)
+            .tenantId(tenantId.value)
+            .clientSecret(clientSecret.value)
+            .build()
+            .getToken((new TokenRequestContext)
+              .setScopes(List("https://storage.azure.com/.default").asJava)))
+          .compile
+          .lastOrError
+          .map(tk => Expires(new TokenCredentials(tk.getToken), tk.getExpiresAt))
     }
 
-  def mkContainerUrl[F[_]](cfg: Config)(implicit F: Sync[F]): F[ContainerURL] =
-    F.catchNonFatal(new URL(cfg.storageUrl.value)) map { url =>
-      val serviceUrl = new ServiceURL(
-        url,
-        StorageURL.createPipeline(
-          mkCredentials(cfg.credentials),
-          // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
-          // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
-          // this results in messages being logged to console in eg quasar's repl
-          // (despite having no console logger configured).
-          // Still not having a full understanding of how this logging is supposed to work
-          // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
-          // `disableDefaultLogging = true` seems to be fixing it.
-          // As an aside, in case we need to look at this in the future,
-          // Azure SDK has 2 loggers for some reason:
-          // - `com.microsoft.azure.storage.blob.LoggingFactory`
-          // - `Azure Storage Java SDK`
-          new PipelineOptions().withLoggingOptions(
-            new LoggingOptions(3000, true))))
-      serviceUrl.createContainerURL(cfg.containerName.value)
+  def mkContainerUrl[F[_]](cfg: Config)(implicit F: ConcurrentEffect[F]): F[Expires[ContainerURL]] =
+    F.catchNonFatal(new URL(cfg.storageUrl.value)) flatMap { url =>
+      // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
+      // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
+      // this results in messages being logged to console in eg quasar's repl
+      // (despite having no console logger configured).
+      // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
+      // `disableDefaultLogging = true` seems to be fixing it.
+      // As an aside, in case we need to look at this in the future,
+      // Azure SDK has 2 loggers for some reason:
+      // - `com.microsoft.azure.storage.blob.LoggingFactory`
+      // - `Azure Storage Java SDK`
+
+      Functor[F].compose[Expires].map(mkCredentials(cfg.credentials)) { creds =>
+        val pipeline =
+          StorageURL.createPipeline(
+            creds, new PipelineOptions().withLoggingOptions(new LoggingOptions(3000, true)))
+
+        new ServiceURL(url, pipeline).createContainerURL(cfg.containerName.value)
+      }
     }
+
+  def refContainerUrl[F[_]: ConcurrentEffect: Timer](cfg: Config)
```

Comment:
Unlike `SharedKey` credentials, AAD credentials expire after a set amount of time. This helper function returns a reference with the `ContainerURL` inside and a function to refresh the token. Whenver the second element of the tuple is sequenced into a computation, the `Ref` is updated with newly acquired credentials.

This is a helper function, implementation are free to use the lower level `mkContainerUrl` if they so choose.

Reply:
Nice!